engine/resources/pippet.go

```diff
@@ -0,0 +1,183 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	//"sync"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+)
+
+func init() {
+	engine.RegisterResource("pippet", func() engine.Res { return &PippetRes{} })
+}
+
+// PippetRes is a wrapper resource for puppet.
+// It implements the functional equivalent of an exec resource
+// that calls "pupept resource <type> <title> <params>", but offers
+// superior performance through a long-running Puppet process that
+// receives resources through a pipe (hence the name).
+type PippetRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Refreshable
+
+	init *engine.Init
+
+	Type   string `yaml:"type" json:"type"`
```

Comment:
Why json?
Also, you'll want to add the `lang` tags only if you want anyone to be able to use this from the DSL. (Do we?)

Reply:
Oops, sorry. That was from an earlier test I did. Removed.

I would not expose this resource type to the language until we get a specific feature request for this. It should be considered an integral part of the Puppet Crutch.